rfcs/0169-feature-parameter-names.md

```diff
@@ -0,0 +1,595 @@
+---
+feature: feature_parameter_names
+start-date: 2024-01-10
+author: Dmitry Bogatov
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Nixpkgs has mechanisms to exports compile-time options of the upstream build
+system through the named parameters of derivation function, but the names of
+these parameters are inconsistent from package to package and from feature to
+feature.
+
+# Motivation
+[motivation]: #motivation
+
+Developer interface for building environment with specific features enabled and
+disabled is more complicated and and involves more "you just have to know it"
+compared to already existing systems.
+
+1. Writing generic code is complicated. Even if list of packages all have
+   optional dependency on package `foo`, duck-typing like
+
+```
+map (p: p.override { with_foo = false; }) packages
+```
+
+won't work because of lack of uniformity, leading to ugly code downstream.
+
+2. If user wants to globally disable support for optional feature or build
+   dependency as much as possible, he needs to set multiple feature parameters,
+   and he still can miss some. E.g, it is not obvious whether the following
+   list is enough to cover everything X11-related:
+
+```
+withX
+withX11
+x11Support
+enableX
+```
+
+3. There is no clear separation between feature parameters and other arguments
+   to the package. With naming mandated by this RFC, distinction is clear to
+   cater to automatic queries.
+
+4. Side effect of the proposdd migration plan is compiling of almost
+   exahustive list of known feature parameters.
+
+
+# Detailed design
+[design]: #detailed-design
+
+When upstream build system has options to enable or disable a optional features
+or configure some compile-time parameter, like path to default mailbox or
```

Comment:
Would this rule apply to feature parameters that do not correspond to upstream build system features?

For exampe, [libiconv](https://github.com/NixOS/nixpkgs/blob/fadc1d7ef25cba8580cc0380f07e686b1ad2fe49/pkgs/development/libraries/libiconv/default.nix#L4) has a feature flag `enableDarwinABICompat` for building libiconv in a way that is ABI-compatible with the system-provided libiconv. The plan is to drop darwin.libiconv in favor of the upstream one, but the upstream one needs this compatibility to allow references to the system one to be replace with it (using `install_name_tool`). Otherwise, doing so will result in crashes due to missing symbols.

Additionally, libiconv (and other packages) feature `enableShared` and `enableStatic` flags. Would these be migrated to the proposed feature flags section or left outside of it?

Reply:
I'd say to migrate these too, in line with suggested `enable_ssl`.

Reply:
Thanks. My other question was about options that are for nixpkgs use. libiconv is the primary one that comes to mind. It builds libivonv in a specific way for ABI compatibility, so it doesn’t map to a specific build system option.

If I am understanding correctly, that type of option would still be okay under this RFC even though it’s not an upstream-provided option. (In this case, we’re manually recreating what Apple does in their fork’s Xcode project.)